ext/AbstractDifferentiationEnzymeExt.jl

```diff
@@ -0,0 +1,60 @@
+module AbstractDifferentiationEnzymeExt
+
+if isdefined(Base, :get_extension)
+    import AbstractDifferentiation as AD
+    using Enzyme: Enzyme
+else
+    import ..AbstractDifferentiation as AD
+    using ..Enzyme: Enzyme
+end
+
+AD.@primitive function jacobian(b::AD.EnzymeForwardBackend, f, x)
+    val = f(x)
+    if val isa Real
+        return adjoint.(AD.gradient(b, f, x))
+    else
+        if length(x) == 1 && length(val) == 1
+            # Enzyme.jacobian returns a vector of length 1 in this case
+            return (Matrix(adjoint(Enzyme.jacobian(Enzyme.Forward, f, x))),)
+        else
+            return (Enzyme.jacobian(Enzyme.Forward, f, x),)
+        end
+    end
+end
+function AD.jacobian(b::AD.EnzymeForwardBackend, f, x::Real)
+    return AD.derivative(b, f, x)
+end
+function AD.gradient(::AD.EnzymeForwardBackend, f, x::AbstractArray)
+    # Enzyme.gradient with Forward returns a tuple of the same length as the input
+    return ([Enzyme.gradient(Enzyme.Forward, f, x)...],)
+end
+function AD.gradient(b::AD.EnzymeForwardBackend, f, x::Real)
+    return AD.derivative(b, f, x)
+end
+function AD.derivative(::AD.EnzymeForwardBackend, f, x::Number)
+    (Enzyme.autodiff(Enzyme.Forward, f, Enzyme.Duplicated(x, one(x)))[1],)
+end
+
+AD.@primitive function jacobian(::AD.EnzymeReverseBackend, f, x)
+    val = f(x)
+    if val isa Real
+        return (adjoint(Enzyme.gradient(Enzyme.Reverse, f, x)),)
```

Comment:
And here

Reply:
why?

Reply:
Moreover for here and all of the above gradient/etc calls a single call to reverse mode AD can take the derivative of an arbitrary number of parameters, which is what the higher level AbstractDifferentiation API provides. In this case if that would force multiple calls to Enzyme, especially in a closure, that would be poor for performance, and likely various GC/type stability.

Reply:
Most tests pass now, a few fail due to potential Enzyme correctness bug. I will resolve this conversation and start another thread for the correctness issues.